src/gt4py/next/allocators.py

```diff
@@ -0,0 +1,303 @@
+    ) -> core_allocators.TensorBuffer[core_defs.DeviceTypeT, core_defs.ScalarT]:
+        raise self.exception
+
+
+if CUPY_DEVICE is not None:
+    cp_alloc_ns: core_allocators.ValidNumPyLikeAllocationNS = cp  # Just for static type checking
+    assert core_allocators.is_valid_nplike_allocation_ns(cp_alloc_ns)
+
+    if CUPY_DEVICE is core_defs.DeviceType.CUDA:
+
+        class CUDAFieldBufferAllocator(BaseFieldBufferAllocator[core_defs.CUDADeviceTyping]):
+            def __init__(self) -> None:
+                super().__init__(
+                    device_type=core_defs.DeviceType.CUDA,
+                    array_ns=cp_alloc_ns,
+                    layout_mapper=horizontal_first_layout_mapper,
+                    byte_alignment=128,
+                )
+
+        device_allocators[core_defs.DeviceType.CUDA] = CUDAFieldBufferAllocator()
+
+    else:
+
+        class ROCMFieldBufferAllocator(BaseFieldBufferAllocator[core_defs.ROCMDeviceTyping]):
+            def __init__(self) -> None:
+                super().__init__(
+                    device_type=core_defs.DeviceType.ROCM,
+                    array_ns=cp_alloc_ns,
+                    layout_mapper=horizontal_first_layout_mapper,
+                    byte_alignment=128,
+                )
+
+        device_allocators[core_defs.DeviceType.ROCM] = ROCMFieldBufferAllocator()
+
+else:
+
+    class InvalidGPUFielBufferAllocator(InvalidFieldBufferAllocator[core_defs.CUDADeviceTyping]):
+        def __init__(self) -> None:
+            super().__init__(
+                device_type=core_defs.DeviceType.CPU,
+                exception=RuntimeError("Missing `cupy` dependency for GPU allocation"),
+            )
+
+
+StandardGPUFieldBufferAllocator: Final[type[FieldBufferAllocatorProtocol]] = (
+    device_allocators[CUPY_DEVICE] if cp else InvalidGPUFielBufferAllocator
+)
+
+
+def allocate(
+    domain: common.DomainLike,
+    dtype: core_defs.DType[core_defs.ScalarT],
+    *,
+    aligned_index: Optional[Sequence[common.NamedIndex]] = None,
+    allocator: Optional[FieldBufferAllocationTool] = None,
+    device: Optional[core_defs.Device] = None,
+) -> core_allocators.TensorBuffer:
+    """
+    Allocate a TensorBuffer with the given settings on the given device.
+
+    The arguments `device` and `allocator` are mutually exclusive.
+    If `device` is specified, the corresponding default allocator
+    (defined in :data:`device_allocators`) is used.
+
+    Args: TODO
+        domain:
+        dtype: Data type descriptor as defined in :meth:`BufferAllocator.allocate`.
+
+        aligned_index: N-dimensional index of the first aligned element as defined
+            in :meth:`BufferAllocator.allocate`.
+    """
+    if device is None and allocator is None:
+        raise ValueError("No 'device' or 'allocator' specified")
+    actual_allocator = get_allocator(allocator, None) if allocator is not None else None
```

Comment:
It seems `get_allocator`'s 2nd argument `default` is never used, but it seems it was introduced to simplify the following code...

Reply:
Fixed